### ISSUE TITLE
Issue with 'New' button on child tab of parent/child frontend edit form with SEF enabled

### ISSUE THREAD

**Issue by pjdevries** [vdm-io/Joomla-Component-Builder#383]
First of all I want to thank  Llewellyn for the time and effort he put into JCB and making it available for free to the community.

Now let me report the first issue I encountered. I have a simple project, with two tables in a parent/child relationship. Following the video's, I created various admin views. The admin view for the parent table has a tab showing the child records, allowing new records to be added and existing records to be edited. Works like a charm :)

I also had JCB create the frontend edit logic. Again following the appropriate video's, I created the necessary custom code to show the associate views in the frontend. Also works like a charm, except for one thing: the **New** button on the child tab does not work as expected. It does with SEF *disabled* but not when SEF is *enabled*. **Edit** links work in both cases though. 

Because the Joomla! router is not exactly my strong suite, I had to spend quite a bit of time trying to figure out what's wrong :) Eventually I traced it back to this piece of code in the `build` method of generated component's `router.php`:
```php
		if (isset($query[`view`]))
		{
			$view = $query[`view`];

			if (empty($query['Itemid']))
			{
				$segments[] = $query[`view`];
			}

			unset($query[`view`]);
		}
```
The second `if` results in the `view` parameter being removed from the resulting uri. The assumption seems to be that the `view` parameter is not needed when the `Itemid` is known. That the Joomla! router can deduce the correct view from that `Itemid`. Apparently that's not the cases in parent/child forms.  

Although having a `view` parameter might be overkill in some or even most cases, I figured it can do no harm when it's always there. So I removed the second `if` resulting the **New** button to work as expected. The other views also seem to do their thing without noticeable issues, 
so this might be a proper solution.

Regards,

Pieter-jan de Vries

**Comment by Llewellynvdm** [comment]
Yes, well here is really a complex issue, I am not sure why it does not work out for you, I can say that on my side I have create new buttons on the front, without the need to change that code, and it works as expected. Here is the little helper function I use for creating edit and create buttons.
> Edit button
```php
	/**
	* Get an edit button
	* 
	* @param  int      $item       The item to edit
	* @param  string   $view       The type of item to edit
	* @param  string   $views      The list view controller name
	* @param  string   $ref        The return path
	* @param  string   $component  The component these views belong to
	* @param  string   $headsup    The message to show on click of button
	*
	* @return  string    On success the full html link
	* 
	*/
	public static function getEditButton(&$item, $view, $views, $ref = '', $component = 'com_[[[component]]]', $headsup = JustTEXT::_('All unsaved work on this page will be lost, are you sure you want to continue?'))
	{
		// get URL
		$url = self::getEditURL($item, $view, $views, $ref, $component);
		// check if we found any
		if (self::checkString($url))
		{
			// get the global settings
			if (!self::checkObject(self::$params))
			{
				self::$params = JComponentHelper::getParams('com_[[[component]]]');
			}
			// get UIKIT version
			$uikit = self::$params->get('uikit_version', 2);
			// check that we have the ID
			if (self::checkObject($item) && isset($item->id))
			{
				// check if the checked_out is available
				if (isset($item->checked_out))
				{
					$checked_out = (int) $item->checked_out;
				}
				else
				{
					$checked_out = self::getVar($view, $item->id, 'id', 'checked_out', '=', str_replace('com_', '', $component));
				}
			}
			elseif (self::checkArray($item) && isset($item['id']))
			{
				// check if the checked_out is available
				if (isset($item['checked_out']))
				{
					$checked_out = (int) $item['checked_out'];
				}
				else
				{
					$checked_out = self::getVar($view, $item['id'], 'id', 'checked_out', '=', str_replace('com_', '', $component));
				}
			}
			elseif (is_numeric($item) && $item > 0)
			{
				$checked_out = self::getVar($view, $item, 'id', 'checked_out', '=', str_replace('com_', '', $component));
			}
			// set the link title
			$title = self::safeString(JText::_('Edit') . ' ' . $view, 'W');
			// check that there is a check message
			if (self::checkString($headsup))
			{
				if (3 == $uikit)
				{
					$href = 'onclick="UIkit.modal.confirm(\''.JText::_($headsup).'\').then( function(){ window.location.href = \'' . $url . '\' } )"  href="javascript:void(0)"';
				}
				else
				{
					$href = 'onclick="UIkit2.modal.confirm(\''.JText::_($headsup).'\', function(){ window.location.href = \'' . $url . '\' })"  href="javascript:void(0)"';
				}
			}
			else
			{
				$href = 'href="' . $url . '"';
			}
			// return UIKIT version 3
			if (3 == $uikit)
			{
				// check if it is checked out
				if (isset($checked_out) && $checked_out > 0)
				{
					// is this user the one who checked it out
					if ($checked_out == JFactory::getUser()->id)
					{
						return ' <a ' . $href . ' uk-icon="icon: lock" title="' . $title . '"></a>';
					}
					return ' <a href="#" disabled uk-icon="icon: lock" title="' . JText::sprintf('% has been checked out by %s', self::safeString($view, 'W'), JFactory::getUser($checked_out)->name) . '"></a>'; 
				}
				// return normal edit link
				return ' <a ' . $href . ' uk-icon="icon: pencil" title="' . $title . '"></a>';
			}
			// check if it is checked out (return UIKIT version 2)
			if (isset($checked_out) && $checked_out > 0)
			{
				// is this user the one who checked it out
				if ($checked_out == JFactory::getUser()->id)
				{
					return ' <a ' . $href . ' class="uk-icon-lock" title="' . $title . '"></a>';
				}
				return ' <a href="#" disabled class="uk-icon-lock" title="' . JText::sprintf('% has been checked out by %s', self::safeString($view, 'W'), JFactory::getUser($checked_out)->name) . '"></a>'; 
			}
			// return normal edit link
			return ' <a ' . $href . ' class="uk-icon-pencil" title="' . $title . '"></a>';
		}
		return '';
	}

	/**
	* Get an edit text button
	* 
	* @param  string   $text       The button text
	* @param  int      $item       The item to edit
	* @param  string   $view       The type of item to edit
	* @param  string   $views      The list view controller name
	* @param  string   $ref        The return path
	* @param  string   $component  The component these views belong to
	* @param  string   $headsup    The message to show on click of button
	*
	* @return  string    On success the full html link
	* 
	*/
	public static function getEditTextButton($text, &$item, $view, $views, $ref = '', $component = 'com_[[[component]]]', $jRoute = true, $class = 'uk-button', $headsup = JustTEXT::_('All unsaved work on this page will be lost, are you sure you want to continue?'))
	{
		// make sure we have text
		if (!self::checkString($text))
		{
			return self::getEditButton($item, $view, $views, $ref, $component, $headsup);
		}
		// get URL
		$url = self::getEditURL($item, $view, $views, $ref, $component, $jRoute);
		// check if we found any
		if (self::checkString($url))
		{
			// get the global settings
			if (!self::checkObject(self::$params))
			{
				self::$params = JComponentHelper::getParams('com_[[[component]]]');
			}
			// get UIKIT version
			$uikit = self::$params->get('uikit_version', 2);
			// check that we have the ID
			if (self::checkObject($item) && isset($item->id))
			{
				// check if the checked_out is available
				if (isset($item->checked_out))
				{
					$checked_out = (int) $item->checked_out;
				}
				else
				{
					$checked_out = self::getVar($view, $item->id, 'id', 'checked_out', '=', str_replace('com_', '', $component));
				}
			}
			elseif (self::checkArray($item) && isset($item['id']))
			{
				// check if the checked_out is available
				if (isset($item['checked_out']))
				{
					$checked_out = (int) $item['checked_out'];
				}
				else
				{
					$checked_out = self::getVar($view, $item['id'], 'id', 'checked_out', '=', str_replace('com_', '', $component));
				}
			}
			elseif (is_numeric($item) && $item > 0)
			{
				$checked_out = self::getVar($view, $item, 'id', 'checked_out', '=', str_replace('com_', '', $component));
			}
			// set the link title
			$title = self::safeString(JText::_('Edit') . ' ' . $view, 'W');
			// check that there is a check message
			if (self::checkString($headsup))
			{
				if (3 == $uikit)
				{
					$href = 'onclick="UIkit.modal.confirm(\''.JText::_($headsup).'\').then( function(){ window.location.href = \'' . $url . '\' } )"  href="javascript:void(0)"';
				}
				else
				{
					$href = 'onclick="UIkit2.modal.confirm(\''.JText::_($headsup).'\', function(){ window.location.href = \'' . $url . '\' })"  href="javascript:void(0)"';
				}
			}
			else
			{
				$href = 'href="' . $url . '"';
			}
			// return UIKIT version 3
			if (3 == $uikit)
			{
				// check if it is checked out
				if (isset($checked_out) && $checked_out > 0)
				{
					// is this user the one who checked it out
					if ($checked_out == JFactory::getUser()->id)
					{
						return ' <a class="' . $class . '" ' . $href . ' title="' . $title . '">' . $text . '</a>';
					}
					return ' <a class="' . $class . '" href="#" disabled title="' . JText::sprintf('% has been checked out by %s', self::safeString($view, 'W'), JFactory::getUser($checked_out)->name) . '">' . $text . '</a>'; 
				}
				// return normal edit link
				return ' <a class="' . $class . '" ' . $href . ' title="' . $title . '">' . $text . '</a>';
			}
			// check if it is checked out (return UIKIT version 2)
			if (isset($checked_out) && $checked_out > 0)
			{
				// is this user the one who checked it out
				if ($checked_out == JFactory::getUser()->id)
				{
					return ' <a class="' . $class . '" ' . $href . ' title="' . $title . '">' . $text . '</a>';
				}
				return ' <a class="' . $class . '" href="#" disabled title="' . JText::sprintf('% has been checked out by %s', self::safeString($view, 'W'), JFactory::getUser($checked_out)->name) . '">' . $text . '</a>'; 
			}
			// return normal edit link
			return ' <a class="' . $class . '" ' . $href . ' title="' . $title . '">' . $text . '</a>';
		}
		return '';
	}

	/**
	* Get the edit URL
	* 
	* @param  int      $item        The item to edit
	* @param  string   $view        The type of item to edit
	* @param  string   $views       The list view controller name
	* @param  string   $ref         The return path
	* @param  string   $component   The component these views belong to
	* @param  bool     $jRoute      The switch to add use JRoute or not
	*
	* @return  string    On success the edit url
	* 
	*/
	public static function getEditURL(&$item, $view, $views, $ref = '', $component = 'com_[[[component]]]', $jRoute = true)
	{
		// make sure the user has access to view
		if (!JFactory::getUser()->authorise($view. '.access', $component))
		{
			return false;
		}
		// build record
		$record = new stdClass();
		// check that we have the ID
		if (self::checkObject($item) && isset($item->id))
		{
			$record->id = (int) $item->id;
			// check if created_by is available
			if (isset($item->created_by) && $item->created_by > 0)
			{
				$record->created_by = (int) $item->created_by;
			}
		}
		elseif (self::checkArray($item) && isset($item['id']))
		{
			$record->id = (int) $item['id'];
			// check if created_by is available
			if (isset($item['created_by']) && $item['created_by'] > 0)
			{
				$record->created_by = (int) $item['created_by'];
			}
		}
		elseif (is_numeric($item))
		{
			$record->id = (int) $item;
		}
		// check ID
		if (isset($record->id) && $record->id > 0)
		{
			// get user action permission to edit
			$action = self::getActions($view, $record, $views, 'edit', str_replace('com_', '', $component));
			// check if the view permission is set
			if (($edit = $action->get($view . '.edit', 'none-set')) === 'none-set')
			{
				// fall back on the core permission then
				$edit = $action->get('core.edit', 'none-set');
			}
			// can edit
			if ($edit)
			{
				// set the edit link
				if ($jRoute)
				{
					return JRoute::_("index.php?option=" . $component . "&view=" . $views . "&task=" . $view . ".edit&id=" . $record->id . $ref);
				}
				return "index.php?option=" . $component . "&view=" . $views . "&task=" . $view . ".edit&id=" . $record->id . $ref;
			}
		}
		return false;
	}
```

Create Button
```php
	/**
	* Get a create button
	*
	* @param  string   $view       The type of item to edit
	* @param  string   $views      The list view controller name
	* @param  string   $ref        The return path
	* @param  string   $component  The component these views belong to
	* @param  string   $headsup    The message to show on click of button
	*
	* @return  string    On success the full html create button
	*
	*/
	public static function getCreateButton($view, $views, $ref = '', $component = 'com_[[[component]]]', $headsup = JustTEXT::_('All unsaved work on this page will be lost, are you sure you want to continue?'))
	{
		// get URL
		$url = self::getCreateURL($view, $views, $ref, $component);
		// check if we found any
		if (self::checkString($url))
		{
			// get the global settings
			if (!self::checkObject(self::$params))
			{
				self::$params = JComponentHelper::getParams('com_[[[component]]]');
			}
			// get UIKIT version
			$uikit = self::$params->get('uikit_version', 2);
			// set the link title
			$title = self::safeString(JText::_('Add') . ' ' . $view, 'W');
			// check that there is a check message
			if (self::checkString($headsup))
			{
				if (3 == $uikit)
				{
					$href = 'onclick="UIkit.modal.confirm(\''.JText::_($headsup).'\').then( function(){ window.location.href = \'' . $url . '\' } )"  href="javascript:void(0)"';
				}
				else
				{
					$href = 'onclick="UIkit2.modal.confirm(\''.JText::_($headsup).'\', function(){ window.location.href = \'' . $url . '\' })"  href="javascript:void(0)"';
				}
			}
			else
			{
				$href = 'href="' . $url . '"';
			}
			// return normal create new link
			return ' <a ' . $href . ' class="btn btn-small button-new btn-success" title="' . $title . '"><span class="icon-new icon-white"></span> ' . JText::_('New') . '</span></a>';
		}
		return '';
	}

	/**
	* Get the create URL
	*
	* @param  string   $view        The type of item to edit
	* @param  string   $views       The list view controller name
	* @param  string   $ref         The return path
	* @param  string   $component   The component these views belong to
	* @param  bool     $jRoute      The switch to add use JRoute or not
	*
	* @return  string    On success the create url
	*
	*/
	public static function  getCreateURL($view, $views, $ref = '', $component = 'com_[[[component]]]', $jRoute = true)
	{
		// can create
		if (JFactory::getUser()->authorise($view. '.access', $component) && JFactory::getUser()->authorise($view . '.create', $component))
		{
			// set create task
			$create = "&task=" . $view . ".edit";
			// check if this button must work with task or layout
			if ($views === $view)
			{
				// set layout edit
				$create = "&layout=edit";
			}
			// set the edit link
			if ($jRoute)
			{
				return JRoute::_("index.php?option=" . $component . "&view=" . $views . $create . $ref);
			}
			return "index.php?option=" . $component . "&view=" . $views . $create . $ref;
		}
		return false;
	}
```

Hope this helps you resolve the issue.

**Comment by pjdevries** [comment]
Thanks for your swift and elaborate reply. Those code snippets may prove quite helpful, but I need some more time to study them.

You mention having created new buttons on the front, without having to change that code. The point is, I did not have to create that new button at all. It's generated by JCB. And the only thing I can say, is that the new button does not work with the generated code, while it does after removing that `if`.

**Comment by Llewellynvdm** [comment]
ahh I get it... okay let me check it out you are not taking about custom scripting... sorry I missed that. Give me a moment and I will investigate.

**Comment by pjdevries** [comment]
Apologies if I was not clear. Not one of my strong points. And do take your time. For the time being I use the custom code insertion feature to work around this :)

**Comment by Llewellynvdm** [comment]
Okay I pushed a fix to the staging branch, please test and let me know how it now works on the frontend.

**Comment by pjdevries** [comment]
Either I'm doing something wrong. or it still does not work. I don't see any changes in the generated router anyway, so I'm curious to learn how you have fixed this.

**Comment by Llewellynvdm** [comment]
Well look at the [code :)](https://github.com/vdm-io/Joomla-Component-Builder/commit/098c82c222f1a79423eaff626504591e5a0a4371) here on [line 8093](https://github.com/vdm-io/Joomla-Component-Builder/commit/098c82c222f1a79423eaff626504591e5a0a4371#diff-001cb31b03c7dade5d5f5df9897ff487R8101) I changed the link being build, in fact I made a few changes in that area to improve the buttons. So if it is not resolved, let me ask did you remove the custom code override?

**Comment by pjdevries** [comment]
I checked the code and although in my installation that change is on a different line, the look the same. And yes, I removed the custom code override. I completely deleted it in JCB and also uninstalled and reinstalled the component completely. The New button on the child tab still takes me from within the edit view back to the main list view.

**Comment by Llewellynvdm** [comment]
I do not understand, please explain more. Since when installing the staging branch of JCB you should find it compiles your component with the changes made to the core, if it does not, then your JCB install is not on the latest changes. The code must change, if not something else is wrong.

**Comment by pjdevries** [comment]
The changes in de the compiler code are there. But the problem is still there as well. The New buttons still does not work without the fix I did myself in router.php.

**Comment by pjdevries** [comment]
I feel I'm taking up too much of your valuable time. Like I mentioned in another issue, this probably has to do with my limited communication skills. Maybe it's better to leave it as it is, since you, or anyone else,  apparently do not experience this problem.

**Comment by ro-ot** [comment]
@pjdevries make no mistake! the front-end edit/create buttons is a pain! Takes a good understanding to code this right. I know @Llewellynvdm has done a tremendous work here. But we need a better answer to this fundamental use case. The controller that handles the permissions is not even in the front-end. No wonder things break!

**Comment by pjdevries** [comment]
I very much respect and appreciate Llewell's work. My suggestions and/or reporting of issues is not intended as negative criticism. Far from that. It's my first attempt to contribute to the Joomla! Open Source community. Unfortunately not that successful apparently.

What I regret most, is not understanding why my proposed, simple solution is not acceptable. On the other hand I understand that Llewell's time is too precious to spend on issues nobody else seems to be experiencing, including himself. So it's better to have him spend his valuable time on important issues and new features.

**Comment by Llewellynvdm** [comment]
Well lets agree that we all are volunteers :+1: and I for one are grateful that guys like you @pjdevries are banging out the implementation edge-cases.

I also actually agree with @ro-ot that these buttons are a pain, and that it needs more attention. I have used workarounds to get this working. But to fix this so it just works out of the box is the ideal.

It is not that your fix does not work for you use-case, it is just that it does not work for many other. So we need to resolve this in a way that work for every use-case. Give me a a few days if not more and I will see if I can improve this.

So @ro-ot do you see an easy and simple solution? please share!

**Comment by pjdevries** [comment]
I have to admit I can not really oversee hoe many use cases there are for which my simple fix does not work. So I wish you guys luck with finding the right solution.

**Comment by Llewellynvdm** [comment]
I take it that this issue was resolved with [4587b5a](https://github.com/tonypartridge/Joomla-Component-Builder/commit/4587b5af5d99037f0222379c09ebff98e8cfcb3d) let me know if you are still having issues :+1:
